src/drivers/gps/gps.cpp

```diff
@@ -232,17 +286,230 @@ GPS::init()
 	return OK;
 }
 
-void
-GPS::task_main_trampoline(void *arg)
+void GPS::task_main_trampoline(void *arg)
 {
 	g_dev->task_main();
 }
 
+int GPS::callback(GPSCallbackType type, void *data1, int data2, void *user)
+{
+	GPS *gps = (GPS *)user;
+
+	switch (type) {
+	case GPSCallbackType::readDeviceData:
+		return gps->pollOrRead((uint8_t *)data1, data2, *((int *)data1));
+
+	case GPSCallbackType::writeDeviceData:
+		return write(gps->_serial_fd, data1, (size_t)data2);
+
+	case GPSCallbackType::setBaudrate:
+		return gps->setBaudrate(data2);
+
+	case GPSCallbackType::gotRTCMMessage:
+		/* not used */
+		break;
+
+	case GPSCallbackType::surveyInStatus:
+		/* not used */
+		break;
+
+	case GPSCallbackType::setClock:
+		px4_clock_settime(CLOCK_REALTIME, (timespec *)data1);
+		break;
+	}
+
+	return 0;
+}
+
+int GPS::pollOrRead(uint8_t *buf, size_t buf_length, int timeout)
+{
+	/* check for new messages. Note that we assume poll_or_read is called with a higher frequency
+	 * than we get new injection messages.
+	 */
+	handleInjectDataTopic();
+
+#ifndef __PX4_QURT
+
+	/* For non QURT, use the usual polling. */
+
+	pollfd fds[1];
+	fds[0].fd = _serial_fd;
+	fds[0].events = POLLIN;
+
+	/* Poll for new data,  */
+	int ret = poll(fds, sizeof(fds) / sizeof(fds[0]), timeout);
+
+	if (ret > 0) {
+		/* if we have new data from GPS, go handle it */
+		if (fds[0].revents & POLLIN) {
+			/*
+			 * We are here because poll says there is some data, so this
+			 * won't block even on a blocking device. But don't read immediately
+			 * by 1-2 bytes, wait for some more data to save expensive read() calls.
+			 * If more bytes are available, we'll go back to poll() again.
+			 */
+			usleep(GPS_WAIT_BEFORE_READ * 1000);
```

Comment:
@bkueng Reducing this will increase CPU load from the GPS app, but will help with the uORB buffer issues you found - how about reducing this by half to re-test?